### DISCUSSION TITLE
Integrate x-cut/k-cut cosmic shear

### DISCUSSION THREAD

**Discussion by pltaylor16** [LSSTDESC/firecrown#351] [Ideas]
We are trying to implement x-cut/k-cut cosmic shear (see e.g. https://arxiv.org/abs/2007.00675) as part of the desc-baryon-challenge. A public version of the code which computes the transformation matrix is available here: https://github.com/pltaylor16/x-cut .

**Comment by vitenti** [comment]
Hi @pltaylor16, do you need guidance on how to incorporate this transformation on C_ells produced by Firecrown? Or are you already creating a PR with this aim?

**Comment by pltaylor16** [comment]
Thanks for the reply @vitenti. Looking for a little guidance on the best way to go about doing this.
Ideally this could be merged into the main branch of Firecrown. 

To give you a little more background, the transformation assumes a fiducial cosmology which is fixed throughout inference. The transformation is then applied to the theory vector, data vector and covariance in a consistent way. One then can choose a new set of angular scale cuts on the transformed data vectors.

**Comment by vitenti** [comment]
Hello @pltaylor16 great, I'll turn this issue into a discussion and once we have a good plan we move to an issue or PR.

**Comment by vitenti** [comment]
Hi @pltaylor16,

I have a few questions about this implementation:

- Is this modification intended for the predicted C^{XY}_ell?
- If so, which XY observables does it apply to?
- Does its applicability depend on any specific parameter?
- Could you clarify the role of the fiducial cosmology? In MCMC analyses, we typically explore both cosmological and analysis-specific parameter spaces. Does this code necessitate a fixed cosmology?

I suspect implementing this might involve adding a new systematic object to modify the TheoryVector computed by TwoPoint. However, clarifying these points will help determine the exact approach.

**Reply by pltaylor16** [reply]
Thanks for the help with this.
1. The transformation is applied to the a) the predicted theory vector, b) the data vector and c) the covariance matrix. 
2. The transformation can be applied to both the C(ell) and \xi(\theta).  
3. The transformation does not depend on any parameters. However to compute the transformation matrix one implicitly assumes a fiducial cosmology. One approach is to compute the transformation matrix externally from firecrown; this might be the easiest?
4. Once the transformation is matrix is computed at a fiducial cosmology we vary the cosmology as usual in the MCMC analysis but keep the transformation fixed. This is permissible since we are essentially just applying a linear transformation to the theory, data and covariance in a consistent way.

**Reply by vitenti** [reply]
No problem!

1. does it need to be applied to all these quantities (a,b,c)? Or any other combination is possible (a), (b), (c), (a, b), and so on? If (a,b,c) is required, then, it would have to take place in the likelihood level (the ConstGaussian object) since only at this stage we have all these elements. 
2. What I meant before was whether X and Y must be both shear or not, and how about correlation between counting and shear?
3. OK, we can include these matrices into the SACC file (which contain data, metadata and calibration data). If you think that we may require recomputing these quantities during a MCMC run, then we should include the computation code here.
4. OK

**Reply by pltaylor16** [reply]
1. All three quantities at once, so (a,b,c).
2. X and Y are both shear. (I suppose if one were to do cross-correlations with CMB lensing then Y could be the CMB kappa map while X is the weak lensing field, but for now this is probably overkill).
4. There is no need to compute these from an MCMC run. They can easily be computed externally.

**Reply by vitenti** [reply]
OK, I think we are almost there. Now suppose that I have a likelihood with X in (s0, s1, l0, l1), where the first two are weak lensing measurements and the last two counts. My theory/data vectors layout may be: 
(Cell^{s0, s0}, Cell^{s1, s1}, Cell^{s0, s1}, Cell^{s0, l0}, Cell^{s1, l1}, Cell^{s0, l1}, Cell^{s1, l0}, Cell^{l0, l0}, Cell^{l1, l1}, Cell^{l0, l1}).

In Firecrown each of these subvectors are objects of the type Statistic, they have a well defined length and the associated theta or ells. Now tell me if I have it right:

- one should apply the transformation to the sub-vector defined by the first three Statistics. 
- we need to extract the right blocks from the covariance matrix (those relating these same first three Statistics) and apply the transformation to it. 
- we do not need to worry about correlations between the first three Statistics and other elements of the theory/data vectors.

If that's correct, I think we can move this back to an issue and open a PR with a draft implementation.

We may also have to open an issue on SACC describing the requirements for saving the necessary metadata/calibration data on a sacc object.

**Reply by pltaylor16** [reply]
That's almost exactly right. The length of the cut data vector will have a well defined length once the scale cuts are chosen in either ell or theta. The choice of ell and theta scale cuts will not match what we would have chosen if we did not make the transformation. We also do need to account for the covariance of the transformed shear Cells with the other Cells. For example if we say that under the transformation C_ell^{s0,s0} --> D_ell^{s0,s0} then Cov(C_ell^{s0,s0}, C_ell^{l1, l1}) --> Cov(D_ell^{s0,s0}, C_ell^{l1, l1}).

**Reply by vitenti** [reply]
OK, since it does not depend on parameters, I assume that the modification is constant during a statistical analysis. If that's the case, we can do the following, we write a function that updates the sacc object (all relevant parts of the covariance and the subvectos of the data vector) and create a systematic object that updates just the subvectors of the theory vector on each iteration. The update on the sacc object can be done by the user previously (and kept saved in a sacc file), or we can include a hook in the likelihood initialization so it can be applied once and before the start of the analysis (if it's not too slow). 

About the ells (or theta). Currently they are described in the sacc object, for each bin of XY you have many data points (C^{XY}_ell, ell) for ell in a set I: 
- after applying the transformation will the set I change?
- is there a restriction on the initial set I?
- what about window functions?

**Reply by pltaylor16** [reply]
-The set of l will not change in the analysis.
-There is no restriction in the initial set l, as long as the user is somehow able to specify what this is at the beginning.
-The window function remains unaltered.

**Reply by vitenti** [reply]
OK, great, I think we have the big picture. Are you available for a hands on meeting with the team so we can try to write a first draft implementation? We meet regularly Tuesdays and Thursdays at 17h GMT, next Tuesdays we are already full, are you available next Thursday? If these dates are not good to you, we can create a whenisgood to schedule a meeting.

**Reply by pltaylor16** [reply]
Fantastic. I am not available next week, but I am available the week after (Jan 22nd -26th). Thurs would be better, but Tuesday is also possible.

**Comment by vitenti** [comment]
Hello @pltaylor16. We have just completed a significant pull request where we implemented new dataclasses to represent the two-point statistics extracted from a SACC object. I propose the following approach for integrating your method:

1. After extracting all two-point data from the SACC file, the resulting dataclasses can be passed to a filtering code, likely incorporating your code (or calling as an external lib).
2. This filter will generate new versions of the data with your method applied. This can be saved or used directly.
3. Since the dataclasses also contain the indices relating their data points to the covariance matrix, the update can be applied there as well.

The only additional code required would be a `Systematic` object to apply the transformation to the theory vector at each step of the statistical analysis.

**Reply by pltaylor16** [reply]
@vitenti Great. This sounds like it will work well. What do you need from me at this stage?

**Reply by vitenti** [reply]
Each measured correlation is stored in a dataclass `TwoPointCell`:
```
class TwoPointCells:
    XY: TwoPointXY
    ells: npt.NDArray[np.int64]
    Cell: None | TwoPointMeasurement = None

class TwoPointXY:
    x: InferredGalaxyZDist
    y: InferredGalaxyZDist
    x_measurement: Measurement
    y_measurement: Measurement

class InferredGalaxyZDist:
    bin_name: str
    z: np.ndarray
    dndz: np.ndarray
    measurements: set[Measurement]

class TwoPointMeasurement:
    data: npt.NDArray[np.float64]
    indices: npt.NDArray[np.int64]
    covariance_name: str

```
Inside the attributes `x` and `y` you can find the n(z) of each tracer, and in `x_measurement` the measurement (for x-cut I suppose we want any pair containing a `Galaxies.SHEAR_E`). The `indices` contain the indices that relate each measurement at `ell` with an entry of the covariance matrix. Finally, the pair `ells` and `data` contain the measured multipoles and their values. 

My idea is to write a filter class/function that will take a set of `TwoPointCells` and a covariance matrix and produce another set with the modifies `ells` and `data` for each `TwoPointCells` and a new covariance matrix. 

My question is:
- Do these classes contain enough information so we can apply you procedure?

Once this is done we need to write a `Systematic` object to apply the appropriate transformation for the theory vector.

---

If all this makes sense, how should we compute the correct transformations? Is there an external library for that, or should we develop something ourselves in Firecrown?

**Reply by pltaylor16** [reply]
All you need to compute the transformation for x-cut is: the n(z) and you need to assume a fiducial cosmology to map z to chi. The transformation is then applied to the C_ell or correlation function. The covariance matrix also needs to be transformed. 

In this repository: https://github.com/pltaylor16/x-cut in BNT.py there is a class to compute the transformation matrix. As you can see it is very straightforward to compute! Applying the transformation to the data vector and covariance will have to be done in firecrown. What do you need from me to do this? I can point you to the equations, or help write the code?

**Reply by vitenti** [reply]
We reworked the TwoPointMeasurement and finally got a final version. I'll try write the code for your transformation and open a PR so you can review it.

**Reply by pltaylor16** [reply]
Great thanks. Let me know if there are issues. One quirk is my code may have a division by zero error if you start at a redshift, z=0. I can't remember if I fixed this.
